### ISSUE TITLE
Roadmap: recurrent candidate-scoring policy, sequence contracts, and value head

### ISSUE THREAD

**Issue by chrismaghuhn** [chrismaghuhn/OCGForge#51]
# Goal

Track the post-Task7 model-architecture program for partial observability: recurrent public-history memory, exact variable-domain candidate scoring, a controlled feed-forward vs recurrent BC comparison, and only later a value head for RL preparation.

Parent roadmap: #16

Phase-6 baseline roadmap: #42

Dataset/scaling roadmap: #43

Phase-7 RL roadmap: #44

This issue is **roadmap/research direction only**. It does not authorize implementation, architecture migration, a new checkpoint, RL, or self-play by itself.

---

# Why this exists

Yu-Gi-Oh! is partially observable. A single current `PlayerObservation` may not contain all policy-relevant public history. A future learned policy may benefit from recurrent memory over perspective-safe public observations and the agent's own prior public semantic actions.

The current candidate-scoring boundary remains mandatory:

```text
PublicEnvironmentObservation_t
+ complete ordered EnvironmentActionCandidate_t[]
→ score every supplied real candidate exactly once
→ choose one existing public_action_key
```

The model never generates legal actions and never repairs, truncates, deduplicates, sorts, or fabricates the environment-supplied domain.

---

# Target architecture hypothesis

Preferred architecture to evaluate after Task7:

```text
PublicEnvironmentObservation_t
        │
        ▼
ObservationEncoder
        │
        ▼
RecurrentCore
h_(t-1) → h_t
        │
        ├───────────────────────────────┐
        │                               │
        ▼                               ▼
ValueHead(h_t)                 EnvironmentActionCandidate_t[]
                                        │
                                        ▼
                                CandidateEncoder
                                        │
                                        ▼
                              e(A0) ... e(A[N-1])
                                        │
                                        ▼
                              score(h_t, e(A_i))
                                        │
                                        ▼
                               exactly N real scores
                                        │
                                        ▼
                              deterministic selection
                                        │
                                        ▼
                           existing public_action_key
```

The current candidate set is an action-domain input to the policy scorer. It is **not** the recurrent memory state itself.

A later recurrent update may also consume the previously selected public semantic action:

```text
h_t = RecurrentCore(
    h_(t-1),
    ObservationEncoder(O_t),
    PreviousActionEncoder(A_(t-1))
)
```

`PreviousActionEncoder` may consume only the accepted public semantic action. It must never consume raw engine response bytes, private response bindings, internal locators, pointers, `CoreHost` state, or hidden information.

---

# Ordering relative to current Phase 6

Do not change Task5, Task6, or Task7 to introduce recurrence or a value head.

Task7 intentionally remains the first strategically meaningful **feed-forward/stateless candidate-scoring BC reference baseline**.

Required ordering:

```text
Task5 trusted evaluation
→ Task6 PyTorch/JAX backend ADR
→ Task7 meaningful feed-forward BC baseline
→ characterize Task7 weaknesses / rare slices
→ recurrent sequence/data contract freeze
→ recurrent candidate-scoring BC implementation
→ controlled feed-forward vs recurrent evaluation
→ architecture decision
→ large-scale dataset growth
→ value-target contract
→ policy + value model
→ supervised value calibration
→ controlled RL comparison
```

This ordering gives a causal reference point. Do not simultaneously change backend, dataset volume, recurrence, value target, RL algorithm, and curriculum then attribute the result to one factor.

---

# R0 — recurrent sequence/data contract freeze

Before recurrent training or large-scale trajectory growth, freeze how trusted episodic trajectories become recurrent training sequences.

At minimum define:

```text
episode_identity
decision_ordinal
observation_t
complete ordered candidate domain_t
chosen public_action_t
previous public semantic action where applicable
terminal/reset boundary
sequence start/end
burn-in semantics
window/truncation semantics
sequence masks / padding semantics
hidden-state initialization/reset rules
```

Required invariants:

- no cross-episode recurrent-state carry;
- no future leakage;
- no hidden/private state in sequence features;
- no chronology loss from shuffling individual decisions;
- no reconstructed candidate domains;
- exact source candidate order retained;
- sequence padding is physical only and never semantic gameplay content;
- any truncated training window retains explicit episode/decision provenance.

A recurrent model must be reproducible from admitted trajectory semantics; recurrent hidden state is derived model state, not new environment authority.

---

# R1 — recurrent candidate-scoring BC

Implement recurrence as a controlled extension of the accepted candidate scorer.

Preserve:

```text
exact same public observation authority
exact same complete candidate domain authority
exact same label/public_action_key meaning
exact same legality boundary
exact same privacy boundary
exact N real candidates → exact N scores
```

The recurrent core summarizes accepted public history. Candidate rows remain independently encoded and scored against `h_t`.

Do not feed internal engine state or an omniscient replay state into recurrence.

---

# R2 — controlled feed-forward vs recurrent comparison

Compare the accepted Task7 feed-forward baseline against the recurrent model under a controlled workload.

Keep fixed where causally possible:

- DatasetManifest / admitted trajectory population;
- train/validation/test episode split;
- candidate encoder and candidate-domain semantics;
- BC objective;
- optimizer/training budget;
- backend/runtime;
- checkpoint/evaluation contracts;
- Task5 frozen evaluation jobs.

Measure at least:

- offline Teacher agreement/loss;
- BC-induced gameplay evaluation;
- first-divergence behavior;
- rare/history-sensitive slices;
- deterministic inference/replay compatibility;
- memory/throughput cost;
- failure/quarantine rate.

Do not assume recurrence wins because the game is partially observable. Adopt it as the default architecture only if evidence supports the additional complexity.

---

# Large-scale data gate

Large-scale trajectory generation should not begin until R0 is accepted and the recurrent experiment has established what sequence/history representation the learner needs.

The purpose is to avoid producing a huge corpus in a storage/training shape that later loses chronology or requires unsafe reconstruction.

Existing immutable trajectories remain authoritative; any recurrent training dataset/view must be a deterministic derived view over admitted episodic data unless a separately reviewed versioned data contract says otherwise.

---

# V0 — value-target contract

The value head is separate from recurrence and must not be introduced in the same causal experiment as the first recurrent policy.

Before adding a value head, freeze a versioned target contract.

A minimal initial terminal-return target may be:

```text
win  = +1
draw =  0
loss = -1
```

but the exact reward/value target, perspective convention, discounting, bootstrap rules, terminal handling, masking, and provenance must be independently specified before implementation.

A learned value is an estimator, not gameplay/reward authority.

---

# V1 — recurrent policy + value model

After the recurrent policy architecture is accepted, add a value head over the recurrent public-history state:

```text
                 h_t
              ┌───┴────┐
              ▼        ▼
Candidate Policy      Value Head
              │        │
       exact N scores  V(h_t)
```

The value head does not consume hidden engine state and does not redefine legal actions.

Initially prefer supervised value calibration on trusted completed trajectories before using the critic inside RL. This provides a sanity check that the recurrent representation can learn meaningful outcome signal without introducing an RL algorithm at the same time.

---

# RL handoff

Controlled RL begins only after:

```text
trusted feed-forward BC reference
+ accepted recurrent sequence/data contract
+ recurrent-vs-feed-forward architecture decision
+ accepted value-target contract
+ accepted policy/value checkpoint path where required
+ trusted Task5 evaluation
```

RL must not be used to repair sequence leakage, hidden-state contamination, candidate-domain truncation, recurrent reset bugs, value-target ambiguity, or checkpoint incompatibility.

---

# Determinism / replay requirements

Recurrent execution adds state and therefore new explicit gates.

At minimum prove:

- exact hidden-state initialization from the declared contract;
- deterministic state transition for identical public sequence + prior recurrent state;
- reset exactly at episode boundaries;
- no dependence on wall time, PID, thread scheduling, worker assignment, filesystem location, or unordered iteration;
- repeated semantic replay yields identical recurrent inputs/state transitions/scores under the declared numeric contract;
- checkpoint identity binds recurrent architecture/configuration;
- hidden recurrent state is never confused with authoritative environment/replay state.

Do not attempt to serialize arbitrary framework-private hidden tensors as gameplay semantic authority.

---

# Privacy requirements

Paired hidden worlds with identical legitimate public histories must produce identical:

```text
sequence inputs
recurrent state under the declared deterministic numeric contract
candidate scores
selected public_action_key
value output once a value head exists
```

No opponent hidden identity, hidden-deck ordering, internal locator, raw response binding, or omniscient debug value may enter recurrent memory directly or indirectly through a hash/cache/key.

---

# Explicit non-goals

This issue does not authorize:

- modifying Task5 evaluation semantics;
- changing the Task6 backend comparison workload;
- replacing the Task7 feed-forward reference baseline;
- large-scale dataset generation before the sequence/data gate;
- RL, self-play, league, MCTS, or search;
- OCGForge-Ignis integration changes;
- a fixed global action vocabulary;
- generated/decoded actions outside the supplied legal candidate domain.

---

# Exit direction

```text
Task7 feed-forward reference accepted
→ R0 recurrent sequence/data contract accepted
→ R1 recurrent BC implemented
→ R2 controlled comparison accepted
→ architecture decision recorded
→ dataset scaling may proceed with sequence requirements known
→ V0 value-target contract
→ V1 policy + value architecture
→ supervised value calibration
→ Phase 7 controlled RL
```

Every transition still requires its own explicit authorization, implementation/research slice, independent live review, and acceptance.

**Comment by chrismaghuhn** [comment]
## Roadmap integration — learned world model / planning

Dedicated planning roadmap: #52

#51 remains the owner of the **recurrent + value + latent hierarchy** architecture. #52 owns the independent learned-world-model / planning extension.

The combined long-term target is now explicitly:

```text
Past    = recurrent memory h_t
Intent  = latent hierarchy z_t + termination β(h,z)
Future  = learned world-model planning
Reality = OCGForge
```

Key boundary:

```text
real h_t != imagined planning state ŝ_k
```

`h_t` is derived only from real perspective-safe accepted history. `ŝ_k` is hypothetical model state and may never become `PlayerObservation`, legality authority, replay authority, semantic gameplay identity, or the next real recurrent state.

#52 also freezes the important future-domain constraint: deep concrete-action MuZero-style search is not authorized merely because a world model exists. At the real root, every searched concrete action is an exact OCGForge `EnvironmentActionCandidate`. At imagined child nodes, future legal domains are unproven unless a later separate contract solves that problem without creating a second rules engine.

The hierarchy and world model remain independently removable. Required controlled comparison eventually includes:

```text
A recurrent + value
B recurrent + value + hierarchy
C recurrent + value + planning
D recurrent + value + hierarchy + planning
```

No current implementation authorization changes. Task5/Task6/Task7 remain unchanged; recurrent, hierarchy, world-model, planning, and RL work still require separate future authorization.

**Comment by chrismaghuhn** [comment]
## Cross-roadmap refinement from #52

The recurrent/hierarchy roadmap should use the stronger trusted-history term from #52:

```text
h_t = REAL-PERSPECTIVE-HISTORY-DERIVED
```

This means `h_t` may include the acting player's legitimately known private information from `PlayerObservation` (for example own hand / own private known cards), while excluding opponent hidden identity/order, FullState-only engine information, omniscient debug state, and every imagined world-model state `ŝ`.

For later hybrid integration:

```text
h_t = real perspective-legal memory/context
z_t = latent strategic skill
ŝ_k = hypothetical planning state only
```

Hierarchy may consume `h_t`; world-model planning may consume a representation derived from `h_t`; imagined `ŝ_k` must never feed back as authoritative recurrent history.

The continuation lock remains unchanged and also constrains future planning integration:

```text
OPTION_TERMINATION_INSIDE_UNFINISHED_CONTINUATION=FORBIDDEN
MANAGER_RESELECTION_INSIDE_UNFINISHED_CONTINUATION=FORBIDDEN
```

#51 remains owner of recurrent memory/value/hierarchy. #52 owns learned-world-model/planning contracts. Neither changes OCGForge legality authority or authorizes implementation.

**Comment by chrismaghuhn** [comment]
## Roadmap amendment — 2026-09-04

This comment refines sequencing/ownership only. The detailed recurrent/value/hierarchy architecture in the issue body remains otherwise unchanged.

### Phase-6 prerequisite update

The old body references a mandatory PyTorch/JAX Task6 bakeoff. That is superseded by the accepted Phase-6 roadmap direction in #16/#42:

```text
Task5 FINAL PASS
→ Task6 PyTorch primary-backend decision + Task7 readiness
→ Task7 meaningful feed-forward/stateless BC reference
→ R0/R1/R2 recurrent program
```

```text
PHASE6_PRIMARY_BACKEND_DIRECTION=PYTORCH
JAX_STATUS=DEFERRED_CANDIDATE
JAX_REJECTED=NO
JAX_PHASE6_REQUIREMENT=NO
NO_BACKEND_BAKEOFF_WITHOUT_MEASURED_REASON=YES
```

Task6 remains a small readiness/decision slice. JAX is not required before this issue's post-Task7 work.

### Real-history recurrent boundary

Use the refined definition:

```text
h_t = REAL-PERSPECTIVE-HISTORY-DERIVED
```

This means accepted information legitimately available to the acting perspective over the real duel history. It is **not** limited to globally public information.

Allowed inputs may include information legitimately known to that player/perspective, including own private known information represented through the accepted perspective-safe observation/history contracts.

Still forbidden:

```text
opponent hidden state
FullState/CoreHost-only facts
private response bindings/internal locators
imagined world-model states
future information
```

No learned belief may be promoted into an authoritative hidden fact.

### Meta-8 placement

Technical curriculum certification is now owned by #55.

Recommended sequencing:

```text
Task7 feed-forward reference
→ #55 Meta-8 technical certification may proceed
→ R0 sequence/data freeze
→ R1 recurrent flat BC
→ R2 feed-forward vs recurrent decision
→ #43 trusted scaling / certified Meta-8 trajectory generation
→ controlled multi-matchup BC
→ V0/V1 value program
→ supervised value calibration
→ controlled recurrent RL baseline
→ H0/H1/H2 hierarchy experiments
```

Meta-8 certification may overlap with R0/R1/R2 planning/implementation because certification is environment/deck trust work, not large-scale training-data generation. Large-scale Meta-8 generation remains gated by both #55 certification and the accepted R0/R2 sequence/architecture decision.

The exact ordering of multi-matchup BC vs V0/V1 can later be frozen experimentally, but do not change curriculum and value/RL architecture in one causal comparison.

### Continuation/hierarchy boundary unchanged

The existing issue rule remains authoritative:

```text
OPTION_TERMINATION_INSIDE_UNFINISHED_CONTINUATION=FORBIDDEN
MANAGER_RESELECTION_INSIDE_UNFINISHED_CONTINUATION=FORBIDDEN
```

Every concrete gameplay action remains one current OCGForge-supplied legal candidate. Latent skills never become deck scripts, macros, or legality authority.

This amendment is roadmap-only and authorizes no R0/R1/R2/V/H implementation.
